packages/istanbul-lib-instrument/src/visitor.js

```diff
@@ -697,9 +697,18 @@ function programVisitor(types, sourceFilePath = 'unknown.js', opts = {}) {
                 .update(JSON.stringify(coverageData))
                 .digest('hex');
             coverageData.hash = hash;
+            const inputSourceMap = coverageData.inputSourceMap;
+            if (inputSourceMap && Object.getPrototypeOf(inputSourceMap) !== Object.prototype) {
+              console.warn("A non-plain object was passed to istanbul-lib-instrument.")
+              const {...inputSourceMapPlain} = inputSourceMap;
+              coverageData.inputSourceMap = inputSourceMapPlain;
+            }
             const coverageNode = T.valueToNode(coverageData);
             delete coverageData[MAGIC_KEY];
             delete coverageData.hash;
+            if (inputSourceMap) {
+              coverageData.inputSourceMap = inputSourceMap;
+            }
```

Comment:
```suggestion
            const inputSourceMap = coverageData.inputSourceMap;
            if (inputSourceMap && Object.getPrototypeOf(inputSourceMap) !== Object.prototype) {
              console.warn("A non-plain object was passed to istanbul-lib-instrument.")
              coverageData.inputSourceMap = {...inputSourceMap}
            }
            const coverageNode = T.valueToNode(coverageData);
            delete coverageData[MAGIC_KEY];
            delete coverageData.hash;
```

If I'm understanding the problem properly, could the implementation be simplified to this ☝️ 

Also, it would be nice to add a test, I think you could just add an extra test case here:

https://github.com/istanbuljs/istanbuljs/blob/c1559005b3bb318da01f505740adb0e782aaf14e/packages/istanbul-lib-instrument/test/specs/input-source-map.yaml#L5

Just add an argument that's something like `inputSourceMapClass: true`.

```
name: defined input source map
code: |
  output = "test"
inputSourceMapNotPojo: { file: "test.js", mappings: "", names: [], sourceRoot: undefined, sources: [ "test.js" ], sourcesContent: [ 'output = "test"' ], version: 3 }
tests:
  - name: sets the input source map
    args: []
    out: "test"
    lines: { '1': 1 }
    statements: { '0': 1 }
    inputSourceMap: { file: "test.js", mappings: "", names: [], sourceRoot: undefined, sources: [ "test.js" ], sourcesContent: [ 'output = "test"' ], version: 3 }
``` 

And in this file, https://github.com/istanbuljs/istanbuljs/blob/c1559005b3bb318da01f505740adb0e782aaf14e/packages/istanbul-lib-instrument/test/specs.test.js#L70

Create a:

```js
class NotPojo {
  constructor(obj) {
    Object.assign(this, obj)
  }
}
```

And create an instance of this if `inputSourceMapNotPojo` is set rather than `inputSourceMap`.

Reply:
Thanks for the suggestions!
- The simplification of the spread operator is a good idea, commited
- In the change suggestion you removed the part where the original sourceMap is restored - I took that over from #576, I am not sure if it's the right approach... Was that on purpose?
- I agree a test would be nice and I tried adding it, but I have no idea how the tests work (I've never used mocha, only jest...). I still have no idea.